Cannot find module '../node_modules/gpu.js/src/index.js'

<!-- If you don't mind add a fun gif or meme, but no pressure -->

![A GIF or MEME to give some spice of the internet](https://preview.redd.it/web-3-devs-turning-into-ai-devs-v0-nx0pi9c9i0aa1.png?width=640&crop=smart&auto=webp&s=2ed1bd4353eb9f9ff137eb08589c5f941d69bb8d)

## _What_ is wrong?

I cannot properly install brain.js on my machine.


## _Where_ does it happen?

Im running on M1 Pro, Node 18.
`yarn add brain.js` adds the `brain.js`, but does not add the `gpu.js` (i.e. there is no gpu.js in `node_modules`).
Adding `gpu.js` manually (i.e. `yarn add gpu.js`) does not help - the same error persists (even though the gpu.js is now in the node_modules):
```
node:internal/modules/cjs/loader:998
  throw err;
  ^

Error: Cannot find module '../node_modules/gpu.js/src/index.js'
Require stack:
- /Users/piotrpedziwiatr/projects/example-redstone/node_modules/brain.js/dist/index.js
- /Users/piotrpedziwiatr/projects/example-redstone/index.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:995:15)
    at Module._load (node:internal/modules/cjs/loader:841:27)
    at Module.require (node:internal/modules/cjs/loader:1061:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/Users/piotrpedziwiatr/projects/example-redstone/node_modules/brain.js/dist/index.js:6:16)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Module.require (node:internal/modules/cjs/loader:1061:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/piotrpedziwiatr/projects/example-redstone/node_modules/brain.js/dist/index.js',
    '/Users/piotrpedziwiatr/projects/example-redstone/index.js'
  ]
}

Node.js v18.12.1
```

The only way to fix it for me is to manually change
`var index_js = require('../node_modules/gpu.js/src/index.js');`
to
`var index_js = require('../../../node_modules/gpu.js/src/index.js');`

on line 6 of the `dist/index.js` in brain.js.


## _How_ do we replicate the issue?

1. Clone https://github.com/warp-contracts/example-redstone/tree/ppe/brain-js-issue
2. `yarn`
3. `node brain-damage.js`

- should end-up with sth like:
<img width="951" alt="image" src="https://user-images.githubusercontent.com/84311757/227789370-2ea9ceed-719b-4a80-8109-9206ff026ea9.png">


## Expected behavior (i.e. solution)

I should be able to import the library in my script without having to manually change
`var index_js = require('../node_modules/gpu.js/src/index.js');`
to
`var index_js = require('../../../node_modules/gpu.js/src/index.js');`

on line 6 of the `node_modules/brain.js/dist/index.js` :-).

## Version information

### Nodejs:
18.12.1

### Browser:
nope

### Brain.js:
2.0.0-beta.20

## _How_ important is this (1-5)?

Well, the basic functionality does not work...so sth like 4? ;-)

## Other Comments

I've just noticed that gpu.js is set as a peerDependency. AFAIK yarn (at least version 1.x) does not install peerDeps by default (in contrary to npm).

I'll perform some more tests using npm

--- a/rollup.config.browser.js
+++ b/rollup.config.browser.js
@@ -2,7 +2,7 @@ import commonjs from '@rollup/plugin-commonjs';
 import json from '@rollup/plugin-json';
 import resolve from '@rollup/plugin-node-resolve';
 import typescript from '@rollup/plugin-typescript';
-import { fileURLToPath } from 'node:url';
+// import { fileURLToPath } from 'node:url';
 
 const name = 'brain';
 const extensions = ['.js', '.json', '.node', '.ts'];
@@ -14,9 +14,13 @@ export default {
   // Specify here external modules which you don't want to include in your bundle (for instance: 'lodash', 'moment' etc.)
   // https://rollupjs.org/guide/en#external-e-external
   external: [
-    fileURLToPath(
-      new URL('./node_modules/gpu.js/src/index.js', import.meta.url)
-    ),
+    // brain js already uses gpu.js as peer dependencies so it shouldn't be like this
+    // fileURLToPath(
+    //   new URL('./node_modules/gpu.js/src/index.js', import.meta.url)
+    // ),
+    
+    // But this
+    'gpu.js'
   ],
 
   plugins: [

--- a/rollup.config.browser.js
+++ b/rollup.config.browser.js
@@ -2,7 +2,7 @@ import commonjs from '@rollup/plugin-commonjs';
 import json from '@rollup/plugin-json';
 import resolve from '@rollup/plugin-node-resolve';
 import typescript from '@rollup/plugin-typescript';
-// import { fileURLToPath } from 'node:url';
+import { fileURLToPath } from 'node:url';
 
 const name = 'brain';
 const extensions = ['.js', '.json', '.node', '.ts'];
@@ -15,12 +15,9 @@ export default {
   // https://rollupjs.org/guide/en#external-e-external
   external: [
     // brain js already uses gpu.js as peer dependencies so it shouldn't be like this
-    // fileURLToPath(
-    //   new URL('./node_modules/gpu.js/src/index.js', import.meta.url)
-    // ),
-    
-    // But this
-    'gpu.js'
+    fileURLToPath(
+      new URL('./node_modules/gpu.js/src/index.js', import.meta.url)
+    ),
   ],
 
   plugins: [

--- a/rollup.config.js
+++ b/rollup.config.js
@@ -3,7 +3,7 @@ import json from '@rollup/plugin-json';
 import resolve from '@rollup/plugin-node-resolve';
 import typescript from '@rollup/plugin-typescript';
 import * as pkg from './package.json';
-import { fileURLToPath } from 'node:url';
+// import { fileURLToPath } from 'node:url';
 
 const extensions = ['.js', '.json', '.node', '.ts'];
 
@@ -13,9 +13,13 @@ export default {
   // Specify here external modules which you don't want to include in your bundle (for instance: 'lodash', 'moment' etc.)
   // https://rollupjs.org/guide/en#external-e-external
   external: [
-    fileURLToPath(
-      new URL('./node_modules/gpu.js/src/index.js', import.meta.url)
-    ),
+    // brain js already uses gpu.js as peer dependencies so it shouldn't be like this
+    // fileURLToPath(
+    //   new URL('./node_modules/gpu.js/src/index.js', import.meta.url)
+    // ),
+    
+    // But this
+    'gpu.js'
   ],
 
   plugins: [

